### PR TITLE
Sensitive locals

### LINKED ISSUE #10436: Suppress provisioner command containing secrets
How can I suppress the echo of the provisioner command that contains secrets passed to it?  A condensed test is as follows:

```
locals {
  secret = aws_secretsmanager("/api_key", null)
  sensitive = true
}

build {
  sources = ["source.amazon-ebs.test"]

  provisioner "ansible-local" {
    command           = "SECRET=${local.secret} PYTHONUNBUFFERED=1 ansible-playbook"
    playbook_file     = "playbook.yml"
```

The ec2 instance running the ansible-local playbook does not have aws creds to retrieve the secret and I need to pass it in.  How can I suppress the echo of `Executing Ansible: cd /tmp/packer-provisioner-ansible-local && SECRET=xxx PYTHONUNBUFFERED=1 ...`

Or asked another way, how can I pass secrets as environment variables into ansible-local and not have them print to stdout?

Thanks!

### PR DESCRIPTION
**Pull request by SwampDragons**
Implements a singular `local` block with a "sensitive" field to allow log filtering for locals. 

Closes #10436

Tested with 

```
variable "party" {
  type = string
  default = "banananas"
}

local "testing" {
  sensitive = true
  expression = "${var.party}"
}

locals {
  local1 = "this"
  local2 = timestamp()
}

source "null" "example" {
  communicator = "none"
}

build {
  sources = [
    "source.null.example"
  ]

  provisioner "shell-local" {
    inline = ["echo testing ${local.testing}",
              "echo testing ${var.party}",
              "echo testing ${local.local1}",
              "echo testing ${local.local2}",
              "echo testing banananas"]    
  }
}
```

### CODE REVIEW

**Review by sylviamoss — APPROVED** [review]
LGTM! 
I would add a local block to the [complete test](https://github.com/hashicorp/packer/blob/master/hcl2template/types.packer_config_test.go#L22) in the `types.packer_config_test.go` just to make sure is parsed along with everything else, and another one in `typers.variables_test.go` in the [basic test](https://github.com/hashicorp/packer/blob/master/hcl2template/types.variables_test.go#L22).

File: hcl2template/types.variables.go [hunk 1]
```diff
@@ -246,6 +249,50 @@ var variableBlockSchema = &hcl.BodySchema{
...
+func decodeLocalBlock(locals []*LocalBlock, block *hcl.Block) (*LocalBlock, hcl.Diagnostics) {
+	for _, local := range locals {
+		if local.Name == block.Labels[0] {
+			return nil, []*hcl.Diagnostic{{
+				Severity: hcl.DiagError,
+				Summary:  "Duplicate variable",
+				Detail:   "Duplicate " + block.Labels[0] + " variable definition found.",
+				Context:  block.DefRange.Ptr(),
+			}}
+		}
```
Comment (azr) [inline hcl2template/types.variables.go:261]: I think this won't detect duplicate variables that are defined in another file. I'd maybe add a test to verify this.

Comment (azr) [reply hcl2template/types.variables.go:261]: As for a suggestion, I'd maybe change the code to be a bit similar to the one for variables:
With the following call : `moreDiags := c.LocalVariables.decodeLocalBlock(block, ectx)` similarly to what in `decodeInputVariables`

File: website/content/docs/templates/hcl_templates/locals.mdx [hunk 2]
```diff
@@ -21,9 +21,17 @@ Guide_](/guides/hcl/variables).
...
 
-Local values are defined in `locals` blocks:
+Local values are defined in `local` or `locals` blocks:
 
 ```hcl
+# Using the local block allows you to mark locals as sensitive, which will
+# filter their values from logs.
+local "mylocal" {
+  expression = "${var.secret_api_key}"
+  sensitive  = true
```
Comment (azr) [inline website/content/docs/templates/hcl_templates/locals.mdx:31]: I like the expression name because that's what it is, it's not only a value 👍🏼

File: hcl2template/types.variables.go (hunk 1 above)
Comment (sylviamoss) [reply hcl2template/types.variables.go:261]: I had the same question in my head but then I realized is checking every existing block of `locals []*LocalBlock`. I tested locally and it catches duplication. Adding a test would be nice to make it more visible.

**Review by azr — COMMENTED** [review]
Nice one LGTM ! Reading the code it looks like this won't detect duplicate local definitions from 2 different files. Other than that this looks super good ! :)

File: hcl2template/types.variables.go (hunk 1 above)
Comment (sylviamoss) [reply hcl2template/types.variables.go:261]: Unless is like in another file 🤔

File: hcl2template/types.packer_config.go [hunk 3]
```diff
@@ -228,7 +232,8 @@ func (c *PackerConfig) evaluateLocalVariable(local *LocalBlock) hcl.Diagnostics
 		return diags
 	}
 	c.LocalVariables[local.Name] = &Variable{
-		Name: local.Name,
+		Name:      local.Name,
+		Sensitive: local.Sensitive,
```
Comment (azr) [inline hcl2template/types.packer_config.go:236]: Here `parseLocalVariables` will be called for each file, and it looks like locals is reset/empty for each new time.

File: hcl2template/types.variables.go (hunk 1 above)
Comment (azr) [reply hcl2template/types.variables.go:261]: yes, `parseLocalVariables(f *hcl.File)` will be called on each hcl file.

Comment (azr) [reply hcl2template/types.variables.go:261]: ( FYI the current code does not look for value duplicate among the HCL locals map because HCL does that for us )

**Comment by SwampDragons** [comment]
I'll add tests and check that it catches duplicates, and if it doesn't I'll make the change you suggested.

File: hcl2template/types.variables.go (hunk 1 above)
Comment (SwampDragons) [reply hcl2template/types.variables.go:261]: Ok you're right, this doesn't catch duplicates in separate files. Reworking.

**Comment by SwampDragons** [comment]
ok, took your feedback and got it to properly check for duplicates across files. Also used the full schema parsing instead of JustAttributes

**Review by sylviamoss — APPROVED** [review]
Nice!

File: hcl2template/types.variables.go [hunk 4]
```diff
@@ -24,6 +24,9 @@ const badIdentifierDetail = "A name must start with a letter or underscore and m
 type LocalBlock struct {
 	Name string
 	Expr hcl.Expression
+	// When Sensitive is set to true Packer will try it best to hide/obfuscate
```
Comment (azr) [inline hcl2template/types.variables.go:27]:
```suggestion
	// When Sensitive is set to true Packer will try it's best to hide/obfuscate
```

**Review by azr — APPROVED** [review]
Super nice one ! LGTM

File: hcl2template/types.packer_config.go [hunk 5]
```diff
@@ -150,10 +150,18 @@ func (c *PackerConfig) parseLocalVariables(f *hcl.File) ([]*LocalBlock, hcl.Diag
 
 	content, moreDiags := f.Body.Content(configSchema)
 	diags = append(diags, moreDiags...)
-	var locals []*LocalBlock
+
+	var locals = c.LocalBlocks
+	if locals == nil {
+		locals = []*LocalBlock{}
+	}
```
Comment (azr) [inline hcl2template/types.packer_config.go:157]: This check is only necessary for maps, the code should work without this.
```suggestion
	locals := c.LocalBlocks
```

**Review by azr — CHANGES_REQUESTED** [review]
Ah, wait I added a duplicate block file for tests and the code panics, this is due do this line: 

https://github.com/hashicorp/packer/blob/a1e9433246659eb721c3ede80247272cf5d7d680/hcl2template/types.packer_config.go#L162-L164

There's an error in the `l, moreDiags := decodeLocalBlock(block, locals) ` call and so we return a nil local block.

One nice way to fix this would be to put the `decodeLocalBlock` as a function to the already existing `Variables` type and then calling `moreDiags := c.LocalVariables.DecodeLocalBlock(block)

**Comment by azr** [comment]
Nice one !
